includes/data/campaign.php

```diff
@@ -1711,9 +1711,18 @@ public function generate_missing_declarations($month_count = 3, $declarations_li
 			// Si il faut ajouter des déclarations
 			if ( !empty( $declarations_limit ) ) {
 				while ( $count_added_declaration < $declarations_limit ) {
-					WDGROIDeclaration::insert( $this->get_api_id(), $current_date->format( 'Y-m-d' ) );
+					// On ne l'ajoute que si elle n'existe pas déjà
+					$add_date = TRUE;
+					foreach ( $existing_roi_declarations as $declaration_object ) {
+						if ( $current_date->format( 'Y-m-d' ) == $declaration_object[ 'date_due' ] ) {
+							$add_date = FALSE;
+						}
+					}
+					if ( $add_date ) {
+						WDGROIDeclaration::insert( $this->get_api_id(), $current_date->format( 'Y-m-d' ) );
+						$count_added_declaration++;
+					}
 					$current_date->add( new DateInterval( 'P'.$month_count.'M' ) );
-					$count_added_declaration++;
 				}

```

Comment:
@emilschn bon, j'ai tenté un algo plus simple d'après ce que j'ai compris. Je n'ai pas testé non plus, tu en penses quoi ?

`	public function generate_missing_declarations($month_count = 3, $declarations_limit = FALSE) {
		// $month_count --> nombre de mois entre 2 déclarations
		// $declarations_limit --> FALSE si on ajoute des échéances "à la main" sans indiquer le nombre
							// --> x si on ajoute x échéances "à la main" en indiquant le nombre
							// --> 1 si c'est une déclaration qui s'ajoute automatiquement à la fin du nb "normal" de déclaration
		// Calcul du nombre de déclarations que devra faire le projet
		$nb_in_a_year = 12 / $month_count;
		$funding_duration = $this->funding_duration();
		if ( $funding_duration == 0 ) {
			$funding_duration = 1;
		}
		$nb_declarations = $funding_duration * $nb_in_a_year;

		// Permet de rajouter des déclarations si nécessaires
		$count_added_declaration = 0;

		if( $declarations_limit == FALSE && ( !isset( $nb_declarations ) || $nb_declarations == 0 ) ) {
			// $declarations_limit --> FALSE si on ajoute des échéances "à la main" sans indiquer le nombre, alors on ajoute le nb de déclarations "normal" qu'on devrait avoir
			$declarations_limit = $nb_declarations;
		}

		if ( !empty( $declarations_limit ) && $declarations_limit > 0 ){
			// Récupération des déclarations existantes
			$existing_roi_declarations = $this->get_roi_declarations();
			// On part de la date de début de versement
			$current_date = new DateTime( $this->first_payment_date() );

			while ( $count_added_declaration < $declarations_limit ) {
				// On ne l'ajoute que si elle n'existe pas déjà
				$add_date = TRUE;
				foreach ( $existing_roi_declarations as $declaration_object ) {
					if ( $current_date->format( 'Y-m-d' ) == $declaration_object[ 'date_due' ] ) {
						$add_date = FALSE;
					}
				}
				if ( $add_date ) {
					WDGROIDeclaration::insert( $this->get_api_id(), $current_date->format( 'Y-m-d' ) );
					$count_added_declaration++;
				}
				$current_date->add( new DateInterval( 'P'.$month_count.'M' ) );
			}
		}
	}`